File: android/app/src/main/java/com/offlinerailmap/android/map/MapMode.kt
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 Joel Haasnoot

package com.offlinerailmap.android.map

import com.offlinerailmap.android.data.Prefs

/**
 * The map "styles" of the OpenRailwayMap web app. Each one is a set of global-state values
 * that the (single) upstream style reacts to; see `knownStyles` in the web UI.
 */
enum class MapMode(val id: String, val label: String, val globalState: Map<String, Any>) {
    STANDARD(
        "standard", "Infrastructure",
        mapOf(
            "tracks" to "usage", "stations" to "station", "pois" to "standard", "turntables" to "plain",
            "platforms" to "plain", "substations" to "none", "boxes" to "none", "catenaries" to "none",
            "switches" to "plain", "signals" to "none",
        ),
    ),
    SPEED(
        "speed", "Speed",
        mapOf(
            "tracks" to "speed", "stations" to "none", "pois" to "none", "turntables" to "none",
            "platforms" to "none", "substations" to "none", "boxes" to "none", "catenaries" to "none",
            "switches" to "none", "signals" to "speed",
        ),
    ),
    SIGNALS(
        "signals", "Train protection",
        mapOf(
            "tracks" to "train_protection", "stations" to "none", "pois" to "signals", "turntables" to "none",
            "platforms" to "none", "substations" to "none", "boxes" to "plain", "catenaries" to "none",
            "switches" to "none", "signals" to "signals",
        ),
    ),
    ELECTRIFICATION(
        "electrification", "Electrification",
        mapOf(
            "tracks" to "electrification", "stations" to "none", "pois" to "electrification", "turntables" to "none",
            "platforms" to "none", "substations" to "plain", "boxes" to "none", "catenaries" to "plain",
            "switches" to "none", "signals" to "electrification",
        ),
    ),
    TRACK(
        "track", "Gauge",
        mapOf(
            "tracks" to "track", "stations" to "none", "pois" to "none", "turntables" to "none",
            "platforms" to "none", "substations" to "none", "boxes" to "none", "catenaries" to "none",
            "switches" to "none", "signals" to "none",
        ),
    ),
    OPERATOR(
        "operator", "Operator",
        mapOf(
            "tracks" to "operator", "stations" to "operator", "pois" to "operator", "turntables" to "none",
            "platforms" to "none", "substations" to "none", "boxes" to "operator", "catenaries" to "none",
            "switches" to "none", "signals" to "none",
        ),
    );

    companion object {
        fun fromId(id: String): MapMode = entries.firstOrNull { it.id == id } ?: STANDARD
    }
}

/** User-configurable options that map onto upstream global-state keys. */
data class MapOptions(
    val showConstruction: Boolean = true,
    val showProposed: Boolean = true,
    val showAbandoned: Boolean = false,
    val showRazed: Boolean = false,
    /** voltageFrequency | maximumCurrent | power */
    val electrificationLine: String = "voltageFrequency",
    /** gauge | loadingGauge | trackClass */
    val trackLine: String = "gauge",
    /** label | name | none */
    val stationLowZoomLabel: String = "label",
) {
    fun save() {
        Prefs.showConstruction = showConstruction
        Prefs.showProposed = showProposed
        Prefs.showAbandoned = showAbandoned
        Prefs.showRazed = showRazed
        Prefs.electrificationLine = electrificationLine
        Prefs.trackLine = trackLine
        Prefs.stationLowZoomLabel = stationLowZoomLabel
    }

    fun toGlobalState(): Map<String, Any> = mapOf(
        "showConstructionInfrastructure" to showConstruction,
        "showProposedInfrastructure" to showProposed,
        "showAbandonedInfrastructure" to showAbandoned,
        "showRazedInfrastructure" to showRazed,
        "electrificationRailwayLine" to electrificationLine,
        "trackRailwayLine" to trackLine,
        "stationLowZoomLabel" to stationLowZoomLabel,
    )

    companion object {
        fun fromPrefs() = MapOptions(
            showConstruction = Prefs.showConstruction,
            showProposed = Prefs.showProposed,
            showAbandoned = Prefs.showAbandoned,
            showRazed = Prefs.showRazed,
            electrificationLine = Prefs.electrificationLine,
            trackLine = Prefs.trackLine,
            stationLowZoomLabel = Prefs.stationLowZoomLabel,
        )
    }
}

File: ios/RailwayMapCore/Sources/RailwayMapCore/MapMode.swift
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 Joel Haasnoot

import Foundation

/**
 The map "styles" of the OpenRailwayMap web app. Each one is a set of global-state values
 that the (single) upstream style reacts to; see `knownStyles` in the web UI.
 */
public enum MapMode: String, CaseIterable, Identifiable, Sendable {
    case standard
    case speed
    case signals
    case electrification
    case track
    case `operator`

    public var id: String {
        rawValue
    }

    public static func fromId(_ id: String) -> MapMode {
        MapMode(rawValue: id) ?? .standard
    }

    public var label: String {
        switch self {
        case .standard:
            return "Infrastructure"
        case .speed:
            return "Speed"
        case .signals:
            return "Train protection"
        case .electrification:
            return "Electrification"
        case .track:
            return "Gauge"
        case .operator:
            return "Operator"
        }
    }

    public var globalState: [String: JSON] {
        switch self {
        case .standard:
            return [
                "tracks": "usage", "stations": "station", "pois": "standard", "turntables": "plain",
                "platforms": "plain", "substations": "none", "boxes": "none", "catenaries": "none",
                "switches": "plain", "signals": "none",
            ]
        case .speed:
            return [
                "tracks": "speed", "stations": "none", "pois": "none", "turntables": "none",
                "platforms": "none", "substations": "none", "boxes": "none", "catenaries": "none",
                "switches": "none", "signals": "speed",
            ]
        case .signals:
            return [
                "tracks": "train_protection", "stations": "none", "pois": "signals", "turntables": "none",
                "platforms": "none", "substations": "none", "boxes": "plain", "catenaries": "none",
                "switches": "none", "signals": "signals",
            ]
        case .electrification:
            return [
                "tracks": "electrification", "stations": "none", "pois": "electrification", "turntables": "none",
                "platforms": "none", "substations": "plain", "boxes": "none", "catenaries": "plain",
                "switches": "none", "signals": "electrification",
            ]
        case .track:
            return [
                "tracks": "track", "stations": "none", "pois": "none", "turntables": "none",
                "platforms": "none", "substations": "none", "boxes": "none", "catenaries": "none",
                "switches": "none", "signals": "none",
            ]
        case .operator:
            return [
                "tracks": "operator", "stations": "operator", "pois": "operator", "turntables": "none",
                "platforms": "none", "substations": "none", "boxes": "operator", "catenaries": "none",
                "switches": "none", "signals": "none",
            ]
        }
    }
}

/// User-configurable options that map onto upstream global-state keys.
public struct MapOptions: Hashable, Sendable {
    public var showConstruction = true
    public var showProposed = true
    public var showAbandoned = false
    public var showRazed = false
    /// voltageFrequency | maximumCurrent | power
    public var electrificationLine = "voltageFrequency"
    /// gauge | loadingGauge | trackClass
    public var trackLine = "gauge"
    /// label | name | none
    public var stationLowZoomLabel = "label"

    public init() {}

    public func toGlobalState() -> [String: JSON] {
        [
            "showConstructionInfrastructure": .bool(showConstruction),
            "showProposedInfrastructure": .bool(showProposed),
            "showAbandonedInfrastructure": .bool(showAbandoned),
            "showRazedInfrastructure": .bool(showRazed),
            "electrificationRailwayLine": .string(electrificationLine),
            "trackRailwayLine": .string(trackLine),
            "stationLowZoomLabel": .string(stationLowZoomLabel),
        ]
    }

    public func save(to prefs: Prefs) {
        prefs.showConstruction = showConstruction
        prefs.showProposed = showProposed
        prefs.showAbandoned = showAbandoned
        prefs.showRazed = showRazed
        prefs.electrificationLine = electrificationLine
        prefs.trackLine = trackLine
        prefs.stationLowZoomLabel = stationLowZoomLabel
    }

    public static func from(_ prefs: Prefs) -> MapOptions {
        var options = MapOptions()
        options.showConstruction = prefs.showConstruction
        options.showProposed = prefs.showProposed
        options.showAbandoned = prefs.showAbandoned
        options.showRazed = prefs.showRazed
        options.electrificationLine = prefs.electrificationLine
        options.trackLine = prefs.trackLine
        options.stationLowZoomLabel = prefs.stationLowZoomLabel
        return options
    }
}
